dbt/adapters/duckdb/impl.py

```diff
@@ -1,18 +1,52 @@
-from dbt.adapters.duckdb import DuckDBConnectionManager
+from dbt.adapters.duckdb.connections import DuckDBConnectionManager
 from dbt.adapters.sql import SQLAdapter
+from dbt.contracts.connection import AdapterResponse
+from dbt.exceptions import InternalException, RuntimeException
 
 
 class DuckDBAdapter(SQLAdapter):
     ConnectionManager = DuckDBConnectionManager
 
     @classmethod
-    def date_function(cls):
+    def date_function(cls) -> str:
         return "now()"
 
     @classmethod
-    def is_cancelable(cls):
+    def is_cancelable(cls) -> bool:
         return False
 
     def valid_incremental_strategies(self):
         """DuckDB does not currently support MERGE statement."""
         return ["append", "delete+insert"]
+
+    def commit_if_has_connection(self) -> None:
+        """This is just a quick-fix. Python models do not execute begin function so the transaction_open is always false."""
+        try:
+            self.connections.commit_if_has_connection()
+        except InternalException:
+            pass
+
+    def submit_python_job(self, parsed_model: dict, compiled_code: str) -> AdapterResponse:
+
+        connection = self.connections.get_if_exists()
+        if not connection:
+            connection = self.connections.get_thread_connection()
+        con = connection.handle._conn
+
+        def load_df_function(table_name: str):
+            """
+            Currently con.table method dos not support fully qualified name - https://github.com/duckdb/duckdb/issues/5038
+
+            Can be replaced by con.table, after it is fixed.
+            """
+            return con.query(f"select * from {table_name}")
+
+        try:
+            exec(compiled_code, {}, {"load_df_function": load_df_function, "con": con})
```

Comment:
great job @tomsej! a completely passer by comment, so feel free to ignore it. 

One piece of feedback we got from dbt labs folks was that they didn't want python models to run in the same process where dbt-core is running, for good reasons.. Looks like @ChenyuLInx had a look at this PR and was not worried about this so maybe my information is outdated.

Reply:
@turbo1912 I think given that DuckDB itself is running in the same process as dbt, that ship has sailed; dbt-duckdb is the 🏴‍☠️ dbt adapter. 😉

Reply:
Yohoho, @turbo1912 :). Think it would be a problem if this adapter would be supported in cloud, but I do not think it will ever be.

Reply:
We also need to run it in the same process, since you can have only one connection to DuckDB